Python modules should be build as type MODULE, not SHARED

https://github.com/google/or-tools/blob/49b6301e1e1e231d654d79b6032e79809868a70e/ortools/init/python/CMakeLists.txt#L6

When shared libraries are built with "-Wl,--no-undefined" to catch any missing link libraries at build time (instead of run time), the build fails when building e.g. `_pywrapinit.so`, e.g. `undefined reference to `PyExc_AttributeError'`.

But _pywrapinit is not a regular shared library, but a dynamically loaded extension module (loaded by either a standalone python or an embedded interpreter). The python symbols provided by the host are not defined at library/module link time, only resolved at runtime. See https://docs.python.org/3/whatsnew/3.8.html#debug-build-uses-the-same-abi-as-release-build
> On the other hand, pkg-config python3.8 --libs no longer contains -lpython3.8. C extensions must not be linked to libpython (except on Android and Cygwin, whose cases are handled by the script); this change is backward incompatible on purpose. (Contributed by Victor Stinner in [bpo-36721](https://bugs.python.org/issue?@action=redirect&bpo=36721).)

When using MODULE, it is possible to specify different linker flags, i.e. use `-DCMAKE_MODULE_LINKER_FLAGS=-Wl,--as-needed` but `-DCMAKE_SHARED_LINKER_FLAGS=-Wl,--as-needed -Wl,--no-undefined`.

```
[  249s] [ 54%] Linking CXX shared library ../../../lib64/_pywrapinit.so
[  249s] cd /home/abuild/rpmbuild/BUILD/or-tools-9.3/build/ortools/init/python && /usr/bin/cmake -E cmake_link_script CMakeFiles/pywrapinit.dir/link.txt --verbose=1
[  249s] /var/lib/build/ccache/bin/c++ -fPIC -O2 -Wall -D_FORTIFY_SOURCE=2 -fstack-protector-strong -funwind-tables -fasynchronous-unwind-tables -fstack-clash-protection -Werror=return-type -flto=auto -g -Wno-error=return-type -DNDEBUG -O2 -g -DNDEBUG -flto=auto -Wl,--as-needed -Wl,--no-undefined -Wl,-z,now -shared -Wl,-soname,_pywrapinit.so -o ../../../lib64/_pywrapinit.so CMakeFiles/pywrapinit.dir/__/__/__/python/ortools/init/initPYTHON_wrap.cxx.o  -Wl,-rpath,"\$ORIGIN:\$ORIGIN/../../ortools/.libs" ../../../lib64/libortools.so.9.3.9999 -ldl /usr/lib64/libz.so /usr/lib64/libabsl_flags_parse.so.2111.0.0 /usr/lib64/libabsl_flags_usage.so.2111.0.0 /usr/lib64/libabsl_flags_usage_internal.so.2111.0.0 /usr/lib64/libabsl_flags.so.2111.0.0 /usr/lib64/libabsl_flags_internal.so.2111.0.0 /usr/lib64/libabsl_flags_marshalling.so.2111.0.0 /usr/lib64/libabsl_flags_reflection.so.2111.0.0 /usr/lib64/libabsl_flags_config.so.2111.0.0 /usr/lib64/libabsl_flags_private_handle_accessor.so.2111.0.0 /usr/lib64/libabsl_flags_commandlineflag.so.2111.0.0 /usr/lib64/libabsl_flags_commandlineflag_internal.so.2111.0.0 /usr/lib64/libabsl_flags_program_name.so.2111.0.0 /usr/lib64/libabsl_random_distributions.so.2111.0.0 /usr/lib64/libabsl_random_seed_sequences.so.2111.0.0 /usr/lib64/libabsl_random_internal_pool_urbg.so.2111.0.0 /usr/lib64/libabsl_random_internal_randen.so.2111.0.0 /usr/lib64/libabsl_random_internal_randen_hwaes.so.2111.0.0 /usr/lib64/libabsl_random_internal_randen_hwaes_impl.so.2111.0.0 /usr/lib64/libabsl_random_internal_randen_slow.so.2111.0.0 /usr/lib64/libabsl_random_internal_platform.so.2111.0.0 /usr/lib64/libabsl_random_internal_seed_material.so.2111.0.0 /usr/lib64/libabsl_random_seed_gen_exception.so.2111.0.0 /usr/lib64/libabsl_raw_hash_set.so.2111.0.0 /usr/lib64/libabsl_hashtablez_sampler.so.2111.0.0 /usr/lib64/libabsl_hash.so.2111.0.0 /usr/lib64/libabsl_city.so.2111.0.0 /usr/lib64/libabsl_low_level_hash.so.2111.0.0 /usr/lib64/libabsl_leak_check.so.2111.0.0 /usr/lib64/libabsl_statusor.so.2111.0.0 /usr/lib64/libabsl_status.so.2111.0.0 /usr/lib64/libabsl_cord.so.2111.0.0 /usr/lib64/libabsl_cordz_info.so.2111.0.0 /usr/lib64/libabsl_cord_internal.so.2111.0.0 /usr/lib64/libabsl_cordz_functions.so.2111.0.0 /usr/lib64/libabsl_exponential_biased.so.2111.0.0 /usr/lib64/libabsl_cordz_handle.so.2111.0.0 /usr/lib64/libabsl_bad_optional_access.so.2111.0.0 /usr/lib64/libabsl_bad_variant_access.so.2111.0.0 /usr/lib64/libabsl_str_format_internal.so.2111.0.0 /usr/lib64/libabsl_synchronization.so.2111.0.0 /usr/lib64/libabsl_stacktrace.so.2111.0.0 /usr/lib64/libabsl_symbolize.so.2111.0.0 /usr/lib64/libabsl_debugging_internal.so.2111.0.0 /usr/lib64/libabsl_demangle_internal.so.2111.0.0 /usr/lib64/libabsl_graphcycles_internal.so.2111.0.0 /usr/lib64/libabsl_malloc_internal.so.2111.0.0 /usr/lib64/libabsl_time.so.2111.0.0 /usr/lib64/libabsl_strings.so.2111.0.0 /usr/lib64/libabsl_strings_internal.so.2111.0.0 /usr/lib64/libabsl_base.so.2111.0.0 /usr/lib64/libabsl_spinlock_wait.so.2111.0.0 /usr/lib64/libabsl_throw_delegate.so.2111.0.0 /usr/lib64/libabsl_int128.so.2111.0.0 /usr/lib64/libabsl_civil_time.so.2111.0.0 /usr/lib64/libabsl_time_zone.so.2111.0.0 /usr/lib64/libabsl_bad_any_cast_impl.so.2111.0.0 /usr/lib64/libabsl_raw_logging_internal.so.2111.0.0 /usr/lib64/libabsl_log_severity.so.2111.0.0 /usr/lib64/libprotobuf.so /usr/lib64/libre2.so.9.0.0 
[  249s] /usr/lib64/gcc/x86_64-suse-linux/11/../../../../x86_64-suse-linux/bin/ld: /tmp/cc5G4Guk.ltrans0.ltrans.o: in function `SWIG_Python_ErrorType(int)':
[  249s] /home/abuild/rpmbuild/BUILD/or-tools-9.3/build/python/ortools/init/initPYTHON_wrap.cxx:876: undefined reference to `PyExc_IOError'
[  249s] /usr/lib64/gcc/x86_64-suse-linux/11/../../../../x86_64-suse-linux/bin/ld: /home/abuild/rpmbuild/BUILD/or-tools-9.3/build/python/ortools/init/initPYTHON_wrap.cxx:873: undefined reference to `PyExc_MemoryError'
[  249s] /usr/lib64/gcc/x86_64-suse-linux/11/../../../../x86_64-suse-linux/bin/ld: /home/abuild/rpmbuild/BUILD/or-tools-9.3/build/python/ortools/init/initPYTHON_wrap.cxx:903: undefined reference to `PyExc_AttributeError'
...
```

--- a/cmake/docs/swig.md
+++ b/cmake/docs/swig.md
@@ -194,7 +194,7 @@ You can use `OUTPUT_DIR` to change the output directory for the `.py` file e.g.:
 
 ```cmake
 swig_add_library(pyFoo
-  TYPE SHARED
+  TYPE MODULE
   LANGUAGE python
   OUTPUT_DIR ${CMAKE_BINARY_DIR}/python/${PROJECT_NAME}/Foo
   SOURCES foo.i)

--- a/ortools/algorithms/python/CMakeLists.txt
+++ b/ortools/algorithms/python/CMakeLists.txt
@@ -3,7 +3,7 @@ set_property(SOURCE knapsack_solver.i PROPERTY SWIG_MODULE_NAME pywrapknapsack_s
 set_property(SOURCE knapsack_solver.i PROPERTY COMPILE_DEFINITIONS
   ${OR_TOOLS_COMPILE_DEFINITIONS} ABSL_MUST_USE_RESULT)
 swig_add_library(pywrapknapsack_solver
-  TYPE SHARED
+  TYPE MODULE
   LANGUAGE python
   OUTPUT_DIR  ${PYTHON_PROJECT_DIR}/algorithms
   SOURCES knapsack_solver.i)

--- a/ortools/constraint_solver/python/CMakeLists.txt
+++ b/ortools/constraint_solver/python/CMakeLists.txt
@@ -4,7 +4,7 @@ set_property(SOURCE routing.i PROPERTY COMPILE_DEFINITIONS
   ${OR_TOOLS_COMPILE_DEFINITIONS} ABSL_MUST_USE_RESULT)
 set_property(SOURCE routing.i PROPERTY COMPILE_OPTIONS -nofastunpack)
 swig_add_library(pywrapcp
-  TYPE SHARED
+  TYPE MODULE
   LANGUAGE python
   OUTPUT_DIR ${PYTHON_PROJECT_DIR}/constraint_solver
   SOURCES routing.i)

--- a/ortools/init/python/CMakeLists.txt
+++ b/ortools/init/python/CMakeLists.txt
@@ -3,7 +3,7 @@ set_property(SOURCE init.i PROPERTY SWIG_MODULE_NAME pywrapinit)
 set_property(SOURCE init.i PROPERTY COMPILE_DEFINITIONS
   ${OR_TOOLS_COMPILE_DEFINITIONS} ABSL_MUST_USE_RESULT)
 swig_add_library(pywrapinit
-  TYPE SHARED
+  TYPE MODULE
   LANGUAGE python
   OUTPUT_DIR  ${PYTHON_PROJECT_DIR}/init
   SOURCES init.i)

--- a/ortools/linear_solver/python/CMakeLists.txt
+++ b/ortools/linear_solver/python/CMakeLists.txt
@@ -3,7 +3,7 @@ set_property(SOURCE linear_solver.i PROPERTY SWIG_MODULE_NAME pywraplp)
 set_property(SOURCE linear_solver.i PROPERTY COMPILE_DEFINITIONS
   ${OR_TOOLS_COMPILE_DEFINITIONS} ABSL_MUST_USE_RESULT)
 swig_add_library(pywraplp
-  TYPE SHARED
+  TYPE MODULE
   LANGUAGE python
   OUTPUT_DIR ${PYTHON_PROJECT_DIR}/linear_solver
   SOURCES linear_solver.i)

--- a/ortools/scheduling/python/CMakeLists.txt
+++ b/ortools/scheduling/python/CMakeLists.txt
@@ -3,7 +3,7 @@ set_property(SOURCE rcpsp.i PROPERTY SWIG_MODULE_NAME pywraprcpsp)
 set_property(SOURCE rcpsp.i PROPERTY COMPILE_DEFINITIONS
   ${OR_TOOLS_COMPILE_DEFINITIONS} ABSL_MUST_USE_RESULT)
 swig_add_library(pywraprcpsp
-  TYPE SHARED
+  TYPE MODULE
   LANGUAGE python
   OUTPUT_DIR ${PYTHON_PROJECT_DIR}/scheduling
   SOURCES rcpsp.i)

